### ISSUE TITLE
Add method matching and delegation debugging aid

### ISSUE THREAD

**Issue by CodingFabian** [raphw/byte-buddy#65]
The biggest problem when using Byte Buddy for me is that it is hard to know what `ElementMatchers` matched what methods and what `MethodDelegation` is applied. While matchers can be inspected easier by printing the matched `MethodDescription`, `MethodDelegation` is trickier to figure out. This can be also seen on numerous *why did my method not get invoked* posts on the mailing list.
`AgentBuilder#Listener` could be a candidate, but I guess a more global *tell me what matched* listener could help.

If a solution exists in Byte Buddy it is too hidden.

**Comment by raphw** [comment]
You are right, this is currently difficult. I will try to think of a way. It is a bit tricky though, to supply better means. Any interceptor could be user-supplied and this approach would require implementors to supply such information.

Idealy, the error messages of any interceptor supply sufficient information for debugging but I am aware that this is currently not always the case.

**Comment by raphw** [comment]
I looked into this and I enriched the error message thrown by `MethodDelegation`. Further than that, I do not think that I want to add any listener abstraction. Normally, it is not possible to recover from a falsly set-up delegation what makes an exception the appropriate reaction.

Also, I went through previous issues on this matter. Many of them suffered from the former visibility abstraction that made methods invisible despite them being visible. This should be solved already.

A detailed explenation of the binding would require user-custom components to provide additional information on a binding failure. I do not want to make this mandatory, though.
